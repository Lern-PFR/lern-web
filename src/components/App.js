function App() {
  return (
    <div className="App">
      <header className="App-header">
        <p>
          Edit <code>src/components/App.js</code> and save to reload.
        </p>
      </header>
    </div>
  );
}

export default App;
