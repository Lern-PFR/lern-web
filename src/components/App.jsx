import { Home } from 'react-feather';
import { MainLayout, StyledDiv } from 'components/shared/layout';
import {
	Canon,
	Trafalgar,
	Paragon,
	GreatPrimer,
	DoublePica,
	BodyCopy,
	Pica,
	LongPrimer,
	Brevier,
	Minion,
} from 'components/shared/typography';
import {
	PrimaryButton,
	StandardButton,
	SubtleButton,
	DropdownButton,
	DangerButton,
	IconButton,
	FloatingActionButton,
} from 'components/shared/buttons';
import {
	elevationFlat,
	elevationFirst,
	elevationSecond,
	elevationThird,
	elevationFourth,
	elevationFifth,
} from 'theme/elevations';
import {
	LabeledCheckbox,
	FieldsetComponent,
	LabeledRadioButton,
	LabelComponent,
	LabeledInput,
} from 'components/shared/form';

/**
 * @name App
 * @description The main component of the application.
 *
 * @author TimothÃ©e Simon-Franza
 */
const App = () => (
	<MainLayout>
		<StyledDiv position="relative">
			<Canon>Canon text element.</Canon>
			<Trafalgar>Trafalgar text element.</Trafalgar>
			<Paragon>Paragon text element.</Paragon>
			<GreatPrimer>GreatPrimer text element.</GreatPrimer>
			<DoublePica>DoublePica text element.</DoublePica>
			<BodyCopy>BodyCopy text element.</BodyCopy>
			<Pica>Pica text element.</Pica>
			<LongPrimer>LongPrimer text element.</LongPrimer>
			<Brevier>Brevier text element.</Brevier>
			<Minion>Minion text element.</Minion>
		</StyledDiv>
		<StyledDiv position="relative">
			<PrimaryButton>Primary button</PrimaryButton>
			<StandardButton>Standard button</StandardButton>
			<SubtleButton>Subtle button</SubtleButton>
			<DropdownButton>Dropdown button</DropdownButton>
			<DangerButton>Danger button</DangerButton>
			<IconButton><Home /></IconButton>
			<FloatingActionButton isIconOnly><Home /></FloatingActionButton>
		</StyledDiv>
		<StyledDiv position="relative" marginTop="50px" display="flex" width="100%">
			<StyledDiv {...elevationFlat} width="260px" height="260px" padding="10px">Flat component</StyledDiv>
			<StyledDiv {...elevationFirst} width="260px" height="260px" padding="10px">5% elevation component</StyledDiv>
			<StyledDiv {...elevationSecond} width="260px" height="260px" padding="10px">10% elevation component</StyledDiv>
			<StyledDiv {...elevationThird} width="260px" height="260px" padding="10px">20% elevation component</StyledDiv>
			<StyledDiv {...elevationFourth} width="260px" height="260px" padding="10px">30% elevation component</StyledDiv>
			<StyledDiv {...elevationFifth} width="260px" height="260px" padding="10px">50% elevation component</StyledDiv>
		</StyledDiv>
		<StyledDiv position="relative" display="flex" flexDirection="column">
			<LabelComponent>Default label component</LabelComponent>
			<LabelComponent textStyle="bodycopy">BodyCopy label component</LabelComponent>
		</StyledDiv>
		<StyledDiv position="relative">
			<FieldsetComponent title="Basic checkbox components">
				<LabeledCheckbox id="lcb5">Labeled checkbox</LabeledCheckbox>
				<LabeledCheckbox labelTextStyle="bodycopy" checked id="lcb6">Default checked labeled checkbox</LabeledCheckbox>
				<LabeledCheckbox disabled id="lcb7">Disabled labeled checkbox</LabeledCheckbox>
				<LabeledCheckbox disabled checked id="lcb8">Disabled default checked labeled checkbox</LabeledCheckbox>
			</FieldsetComponent>
		</StyledDiv>
		<StyledDiv position="relative">
			<FieldsetComponent title="Basic radio button components">
				<LabeledRadioButton name="rg2" id="lrb5">Labeled radio button</LabeledRadioButton>
				<LabeledRadioButton labelTextStyle="bodycopy" name="rg2" checked id="lrb6">Default checked labeled radio button</LabeledRadioButton>
				<LabeledRadioButton name="rg2" disabled id="lrb7">Disabled labeled radio button</LabeledRadioButton>
			</FieldsetComponent>
		</StyledDiv>
		<StyledDiv position="relative" width="500px">
			<FieldsetComponent title="Text input components">
				<LabeledInput id="basicTextInput">Labeled text input</LabeledInput>
				<LabeledInput id="placeholderTextInput" placeholder="Enter details here.">Labeled text input with custom placeholder</LabeledInput>
				<LabeledInput id="hintTextInput" placeholder="Enter details here." hintText="This is a hint for the user.">Labeled text input with hint</LabeledInput>
				<LabeledInput
					id="disabledTextInput"
					placeholder="Enter details here."
					hintText="This is a hint for the user."
					disabled
				>
					Disabled labeled text input
				</LabeledInput>
				<LabeledInput
					id="errorTextInput"
					placeholder="Enter details here."
					hintText="This is a hint for the user."
					errorText="There is something wrong with the value in this input"
					error
				>
					Labeled text input with a validation error
				</LabeledInput>
				<LabeledInput id="numberInput" placeholder="Number Input" type="number">Labeled number input</LabeledInput>
				<LabeledInput id="passwordInput" placeholder="Password Input" type="password">Labeled password input</LabeledInput>
			</FieldsetComponent>
		</StyledDiv>
	</MainLayout>
);

export default App;
