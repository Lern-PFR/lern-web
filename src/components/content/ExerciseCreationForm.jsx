import { useCallback, useRef, useState } from 'react';
import PropTypes from 'prop-types';
import { useTranslation } from 'react-i18next';

import { isFormValid, validateField, validateForm } from 'lib/shared/formUtils';
import { hasMaxLength, hasMinLength, isRequired } from 'lib/shared/inputValidation';

import { PrimaryButton } from 'components/shared/buttons';
import { LabeledTextArea } from 'components/shared/form';
import { StyledDiv, StyledForm } from 'components/shared/styledElements';

import { form, formSubmit, responseDiv } from 'theme/contentEditionCommon/genericLayout';
import { GreatPrimer } from 'components/shared/typography';
import ResponseItem from './ResponseItem';

/**
 * @constant
 * @name initialState
 * @description The form's initial state value.
 *
 * @author Timothée Simon-Franza
 */
const initialState = {
	title: '',
	description: '',
	content: '',
};

/**
 * @constant
 * @name validationRules
 * @description The validation rules for each field.
 *
 * @author Timothée Simon-Franza
 */
const validationRules = {
	instructions: {
		required: isRequired('required'),
		hasMinLength: hasMinLength(10, 'min_length'),
		hasMaxLength: hasMaxLength(300, 'max_length'),
	},
	explanation: {
		hasMaxLength: hasMaxLength(300, 'max_length'),
	},
};

/**
 * @constant
 * @name inputsDefinition
 * @description The form's input definition object.
 *
 * @author Christopher Walker
 */
const inputsDefinition = {
	instructions: { id: 'instructions', name: 'instructions', inputType: 'textarea', hasPlaceholder: true },
	explanation: { id: 'explanation', name: 'explanation', inputType: 'textarea', hasPlaceholder: true },
};

/**
 * @constant
 * @name inputsDefinition
 * @description The form's input definition object.
 *
 * @author Christopher Walker
 */
const responseFieldsDefinition = {
	answer1: { id: 'answer1', name: 'answer1', inputType: 'responseItem', hasPlaceholder: true, isCorrect: true },
	answer2: { id: 'answer2', name: 'answer2', inputType: 'responseItem', hasPlaceholder: true, isCorrect: false },
	answer3: { id: 'answer3', name: 'answer3', inputType: 'responseItem', hasPlaceholder: true, isCorrect: false },
};

/**
 * @name ExerciseCreationForm
 * @description A form component used to create a lesson.
 *
 * @author Christopher Walker
 *
 * @param {func} onSubmit The method to trigger on form submit.
 */
const ExerciseCreationForm = ({ onSubmit }) => {
	const { t } = useTranslation();
	const [formState, setFormState] = useState(initialState);
	const [errors, setErrors] = useState({});
	const fieldsRef = useRef({});
	const [validAnswer, setValidAnswer] = useState('answer1');

	const hasFormBeenSubmitted = useRef(false);

	/**
	 * @function
	 * @name handleSubmit
	 * @description Handles the form's submit event, performs validation and calls the onSubmit method if all is in order.
	 *
	 * @author Timothée Simon-Franza
	 *
	 * @param {object} event	The event originating from the form's submission.
	 */
	const handleSubmit = useCallback((event) => {
		event?.preventDefault();

		hasFormBeenSubmitted.current = true;
		const formValidationResult = validateForm(fieldsRef.current, validationRules);
		setErrors(formValidationResult);

		if (!isFormValid(formValidationResult)) {
			return;
		}
		const result = {
			statement: formState.instructions,
			explanation: formState.explanation,
			type: 'SingleChoice',
			answers: [
				{
					text: formState.answer1,
					valid: validAnswer === 'answer1',
				},
				{
					text: formState.answer2,
					valid: validAnswer === 'answer2',
				},
				{
					text: formState.answer3,
					valid: validAnswer === 'answer3',
				},
			],
		};
		onSubmit({ ...result });
	}, [formState, onSubmit, validAnswer]);

	/**
	 * @function
	 * @name handleChange
	 * @description Updates the formState object with the new value.
	 *
	 * @author Timothée Simon-Franza
	 *
	 * @param {object} event				The onChange event.
	 * @param {object} event.target			The target which the onChange event originated from.
	 * @param {string} event.target.name	The name of the input the event originated from.
	 * @param {string} event.target.value	The new value to update the state with.
	 */
	const handleChange = useCallback((event) => {
		event.preventDefault();
		const { target: { name, value: newValue } } = event;

		setFormState((prevState) => ({ ...prevState, [name]: newValue }));

		// Prevents the validation error message to be displayed while the user types in for the first time.
		if (hasFormBeenSubmitted.current || errors[name]) {
			const { [name]: removedError, ...rest } = errors; 								// Removes previous errors for current field.
			setErrors({ ...rest, [name]: validateField(newValue, validationRules[name]) });	// Performs validation check on the field and updates the errors state.
		}
	}, [errors, setErrors, setFormState]);

	const handleValidAnswerChange = useCallback((name) => {
		setValidAnswer(name);
	}, [setValidAnswer]);

	/**
	 * @function
	 * @name handleBlur
	 * @description Performs validation checks for the input generating the blur event.
	 *
	 * @author Timothée Simon-Franza
	 *
	 * @param {object} event				The onBlur event.
	 * @param {object} event.target			The target which the onBlur event originated from.
	 * @param {string} event.target.name	The name of the input the event originated from.
	 * @param {string} event.target.value	The current value of the input the event originated from.
	 */
	const handleBlur = useCallback((event) => {
		const { target: { name, value } } = event;

		// Removes previous errors for current field.
		const { [name]: removedError, ...rest } = errors;

		// Performs validation check on the field and updates the errors state.
		setErrors({ ...rest, [name]: validateField(value, validationRules[name]) });
	}, [errors, setErrors]);

	/**
	 *
	 * @param {*} fieldName
	 * @returns
	 */
	const getErrorMessageByFieldName = (fieldName) => {
		if (errors[fieldName] && Object.keys(errors[fieldName])?.length > 0) {
			return t(`exercises.creation.form.fields.${fieldName}.validation_rules.${Object.values(errors[fieldName])?.[0]}`);
		}

		return '';
	};

	return (
		<StyledForm onSubmit={handleSubmit} {...form}>
			{Object.values(inputsDefinition).map(({ id, name, inputType = 'text', hasPlaceholder }) => (
				<LabeledTextArea
					key={id}
					id={id}
					name={name}
					type={inputType}
					onChange={handleChange}
					onBlur={handleBlur}
					rows={6}
					hasError={errors[name] && Object.keys(errors[name])?.length > 0}
					errorText={getErrorMessageByFieldName(name)}
					placeholder={hasPlaceholder ? t(`exercises.creation.form.fields.${name}.placeholder`) : undefined}
					ref={(fieldRef) => { fieldsRef.current[name] = fieldRef; }}
					defaultValue={formState[name]}
				>
					{t(`exercises.creation.form.fields.${name}.label`)}
				</LabeledTextArea>
			))}
			<StyledDiv {...responseDiv}>
				<GreatPrimer>{t('exercises.creation.responses')}</GreatPrimer>
				{Object.values(responseFieldsDefinition).map(({ id, name, hasPlaceholder, isCorrect }) => (
					<ResponseItem
						key={id}
						id={id}
						name={name}
						hasPlaceholder={hasPlaceholder}
						handleCheckedChanged={handleValidAnswerChange}
						handleChange={handleChange}
						handleBlur={handleBlur}
						hasError={errors[name] && Object.keys(errors[name])?.length > 0}
						errorText={getErrorMessageByFieldName(name)}
						checked={isCorrect}
					/>
				))}
			</StyledDiv>
			<PrimaryButton type="submit" {...formSubmit}>
				{t('exercises.creation.form.action.submit')}
			</PrimaryButton>
		</StyledForm>
	);
};

ExerciseCreationForm.propTypes = {
	onSubmit: PropTypes.func.isRequired,
};

export default ExerciseCreationForm;
