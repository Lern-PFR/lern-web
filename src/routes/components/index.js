import RenderRoutes from './RenderRoutes';

export {
	// TODO: disable this exception once there is more to export
	// eslint-disable-next-line import/prefer-default-export
	RenderRoutes,
};
