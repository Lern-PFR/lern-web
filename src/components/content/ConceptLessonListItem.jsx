import { useCallback, useMemo } from 'react';
import { useDispatch } from 'react-redux';
import PropTypes from 'prop-types';
import { generatePath } from 'react-router';
import styled from 'styled-components';
import { useTranslation } from 'react-i18next';

import routes from 'routes';
import { history } from 'routes/components/RouterProvider';

import { StyledDiv, StyledListItem } from 'components/shared/styledElements';
import { BodyCopy, Brevier } from 'components/shared/typography';
import { OutlinedButton, OutlinedLinkButton } from 'components/shared/buttons';
import Link from 'components/shared/navigation/Link';

import { innerContentListItem, innerContentListItemActionsRow, moduleCard } from 'theme/contentEditionCommon/innerContentListTheme';
import { deleteLesson } from 'redux/actions/lessons';

const ListCard = styled(StyledListItem)(
	{
		...innerContentListItem,
		'&, & *': {
			cursor: 'pointer',
		},
	}
);

/**
 * @name ConceptLessonListItem
 * @description A component used to display a concept's lesson in the concept edition page.
 *
 * @author TimothÃ©e Simon-Franza
 *
 * @param {string} description	The current lesson's description.
 * @param {string} id			The current lesson's id.
 * @param {string} title		THe current lesson's title.
 */
const ConceptLessonListItem = ({ description, id, title }) => {
	const dispatch = useDispatch();
	const { t } = useTranslation();

	const editPageRedirectionLink = useMemo(() => generatePath(routes.lessons.lessonEdition, { lessonId: id }), [id]);

	const onCardClick = useCallback(() => {
		history.push(editPageRedirectionLink);
	}, [editPageRedirectionLink]);

	const onDeleteBtnClick = useCallback((e) => {
		if (e?.stopPropagation) {
			e.stopPropagation();
		}

		dispatch(deleteLesson(id));
	}, [dispatch, id]);

	return (
		<ListCard key={id} onClick={onCardClick} {...moduleCard}>
			<BodyCopy margin="0">{title}</BodyCopy>
			<Brevier margin="0">{description}</Brevier>
			<StyledDiv {...innerContentListItemActionsRow}>
				<OutlinedLinkButton>
					<Link to={editPageRedirectionLink}>{t('concepts.edition.lessons_list.links.edit')}</Link>
				</OutlinedLinkButton>
				<OutlinedButton onClick={onDeleteBtnClick}>
					{t('concepts.edition.lessons_list.links.delete')}
				</OutlinedButton>
			</StyledDiv>
		</ListCard>
	);
};

ConceptLessonListItem.propTypes = {
	description: PropTypes.string.isRequired,
	id: PropTypes.string.isRequired,
	title: PropTypes.string.isRequired,
};

export default ConceptLessonListItem;
