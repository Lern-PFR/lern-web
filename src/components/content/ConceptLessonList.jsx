import { useCallback } from 'react';
import { useTranslation } from 'react-i18next';
import PropTypes from 'prop-types';
import { PlusCircle } from 'react-feather';
import { generatePath } from 'react-router';
import styled from 'styled-components';

import routes from 'routes';
import { StyledDiv, StyledList, StyledListItem } from 'components/shared/styledElements';
import { DoublePica } from 'components/shared/typography';
import { history } from 'routes/components/RouterProvider';

import { addInnerContentCard, innerContentList, innerContentListItem } from 'theme/contentEditionCommon/innerContentListTheme';
import ConceptLessonListItem from './ConceptLessonListItem';

const ListCard = styled(StyledListItem)(
	{
		...innerContentListItem,
	}
);

/**
 * @name ConceptLessonList
 * @description A component used in the concept edition page to display its lessons and a "create new lesson" link.
 *
 * @author TimothÃ©e Simon-Franza
 *
 * @param {Array}	lessonList	The lesson list of the concept currently being edited.
 * @param {string}	conceptId	The id of the parent concept.
 */
const ConceptLessonList = ({ lessonList, conceptId }) => {
	const { t } = useTranslation();

	const onAddLessonCardClick = useCallback(() => {
		history.push(generatePath(routes.concepts.lessonCreation, { conceptId }));
	}, [conceptId]);

	return (
		<StyledDiv marginTop="64px">
			<DoublePica as="h3">{t('concepts.edition.lessons_list.title')}</DoublePica>
			<StyledList {...innerContentList}>
				{lessonList.map((lesson) => (
					<ConceptLessonListItem key={lesson.id} {...lesson} />
				))}
				<ListCard {...addInnerContentCard} onClick={onAddLessonCardClick}>
					<PlusCircle />
					{t('concepts.edition.lessons_list.links.add')}
				</ListCard>
			</StyledList>
		</StyledDiv>
	);
};

ConceptLessonList.propTypes = {
	lessonList: PropTypes.arrayOf(
		PropTypes.shape({
			description: PropTypes.string.isRequired,
			id: PropTypes.string.isRequired,
			title: PropTypes.string.isRequired,
		}),
	).isRequired,
	conceptId: PropTypes.string.isRequired,
};

export default ConceptLessonList;
