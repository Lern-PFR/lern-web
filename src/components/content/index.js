import ConceptCreationForm from './ConceptCreationForm';
import ConceptEditionForm from './ConceptEditionForm';
import ConceptLessonList from './ConceptLessonList';
import ModuleCreationForm from './ModuleCreationForm';
import ModuleEditionForm from './ModuleEditionForm';
import ModuleConceptList from './ModuleConceptList';
import NavigationSidebar from './NavigationSidebar';
import SubjectCreationForm from './SubjectCreationForm';
import SubjectEditionForm from './SubjectEditionForm';
import SubjectModuleList from './SubjectModuleList';

export {
	ConceptCreationForm,
	ConceptEditionForm,
	ConceptLessonList,
	ModuleCreationForm,
	ModuleEditionForm,
	ModuleConceptList,
	NavigationSidebar,
	SubjectCreationForm,
	SubjectEditionForm,
	SubjectModuleList,
};
