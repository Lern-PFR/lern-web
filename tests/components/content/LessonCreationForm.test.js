import { shallow } from 'enzyme';
import { LessonCreationForm } from 'components/content';
import { act, fireEvent, render } from '@testing-library/react';

describe('Concept creation form', () => {
	afterEach(() => {
		jest.clearAllMocks();
		jest.restoreAllMocks();
	});

	describe('Snapshot testing', () => {
		const wrapper = shallow(<LessonCreationForm onSubmit={jest.fn()} />);

		expect(wrapper).toMatchSnapshot();
	});

	describe('Form inputs', () => {
		it('should contain a "title" text input.', () => {
			const { container } = render(<LessonCreationForm onSubmit={jest.fn()} />);
			expect(container.querySelector('input[name="title"]')).not.toEqual(null);
		});

		it('should contain a "description" textarea.', () => {
			const { container } = render(<LessonCreationForm onSubmit={jest.fn()} />);
			expect(container.querySelector('textarea[name="description"]')).not.toEqual(null);
		});
	});

	describe('Form inputs validation', () => {
		describe('title input validation', () => {
			let titleInput;
			let submitButton;
			let container;
			let mockedOnSubmit;

			beforeEach(() => {
				mockedOnSubmit = jest.fn();
				container = render(<LessonCreationForm onSubmit={mockedOnSubmit} />).container;
				titleInput = container.querySelector('input[name="title"]');
				submitButton = container.querySelector('button[type="submit"]');
			});

			describe('on submit validation', () => {
				it('should trigger a "required" validation error on the title input if it is empty on submit', () => {
					act(() => {
						fireEvent.input(titleInput, { target: { value: '' } });
						fireEvent.click(submitButton);
					});

					expect(mockedOnSubmit).not.toHaveBeenCalled();
					expect(container.textContent).toMatch(new RegExp('lessons.creation.form.fields.title.validation_rules.required'));
				});

				it('should trigger a "min_length" validation error on the title input if it has less than 3 characters on submit', () => {
					act(() => {
						fireEvent.input(titleInput, { target: { value: 'a' } });
						fireEvent.click(submitButton);
					});

					expect(mockedOnSubmit).not.toHaveBeenCalled();
					expect(container.textContent).toMatch(new RegExp('lessons.creation.form.fields.title.validation_rules.min_length'));
				});

				it('should trigger a "max_length" validation error on the title input if it has more than 50 characters on submit', () => {
					act(() => {
						fireEvent.input(titleInput, { target: { value: 'a'.repeat(51) } });
						fireEvent.click(submitButton);
					});

					expect(mockedOnSubmit).not.toHaveBeenCalled();
					expect(container.textContent).toMatch(new RegExp('lessons.creation.form.fields.title.validation_rules.max_length'));
				});
			});

			describe('on change validation', () => {
				// We only check the requried validation rule since all validation rules are checked.
				it('should trigger a "required" validation error on the title input on change after the first submit', () => {
					act(() => {
						fireEvent.input(titleInput, { target: { value: '' } });
						expect(container.textContent).not.toMatch(new RegExp('lessons.creation.form.fields.title.validation_rules.required'));
						fireEvent.click(submitButton);
						fireEvent.input(titleInput, { target: { value: 'a'.repeat(10) } });
						expect(container.textContent).not.toMatch(new RegExp('lessons.creation.form.fields.title.validation_rules.required'));

						fireEvent.input(titleInput, { target: { value: '' } });
					});

					expect(container.textContent).toMatch(new RegExp('lessons.creation.form.fields.title.validation_rules.required'));
				});
			});
		});

		describe('description textarea validation', () => {
			let descriptionTextarea;
			let submitButton;
			let container;
			let mockedOnSubmit;

			beforeEach(() => {
				mockedOnSubmit = jest.fn();
				container = render(<LessonCreationForm onSubmit={mockedOnSubmit} />).container;
				descriptionTextarea = container.querySelector('textarea[name="description"]');
				submitButton = container.querySelector('button[type="submit"]');
			});

			describe('on submit validation', () => {
				it('should trigger a "required" validation error on the description textarea if it is empty on submit', () => {
					act(() => {
						fireEvent.input(descriptionTextarea, { target: { value: '' } });
						fireEvent.click(submitButton);
					});

					expect(mockedOnSubmit).not.toHaveBeenCalled();
					expect(container.textContent).toMatch(new RegExp('lessons.creation.form.fields.description.validation_rules.required'));
				});

				it('should trigger a "min_length" validation error on the description textarea if it has less than 10 characters on submit', () => {
					act(() => {
						fireEvent.input(descriptionTextarea, { target: { value: 'a' } });
						fireEvent.click(submitButton);
					});

					expect(mockedOnSubmit).not.toHaveBeenCalled();
					expect(container.textContent).toMatch(new RegExp('lessons.creation.form.fields.description.validation_rules.min_length'));
				});

				it('should trigger a "max_length" validation error on the description textarea if it has more than 301 characters on submit', () => {
					act(() => {
						fireEvent.input(descriptionTextarea, { target: { value: 'a'.repeat(301) } });
						fireEvent.click(submitButton);
					});

					expect(mockedOnSubmit).not.toHaveBeenCalled();
					expect(container.textContent).toMatch(new RegExp('lessons.creation.form.fields.description.validation_rules.max_length'));
				});
			});

			describe('on change validation', () => {
				// We only check the requried validation rule since all validation rules are checked.
				it('should trigger a "required" validation error on the description input on change after the first submit', () => {
					act(() => {
						fireEvent.input(descriptionTextarea, { target: { value: '' } });
						expect(container.textContent).not.toMatch(new RegExp('lessons.creation.form.fields.description.validation_rules.required'));
						fireEvent.click(submitButton);
						fireEvent.input(descriptionTextarea, { target: { value: 'a'.repeat(10) } });
						expect(container.textContent).not.toMatch(new RegExp('lessons.creation.form.fields.description.validation_rules.required'));

						fireEvent.input(descriptionTextarea, { target: { value: '' } });
					});

					expect(container.textContent).toMatch(new RegExp('lessons.creation.form.fields.description.validation_rules.required'));
				});
			});
		});
	});

	/*
	describe('content textarea validation', () => {
		let contentTextarea;
		let submitButton;
		let container;
		let mockedOnSubmit;

		beforeEach(() => {
			mockedOnSubmit = jest.fn();
			container = render(<LessonCreationForm onSubmit={mockedOnSubmit} />).container;
			contentTextarea = container.querySelector('textarea[name="content"]');
			submitButton = container.querySelector('button[type="submit"]');
		});

		describe('on submit validation', () => {
			it('should trigger a "required" validation error on the content textarea if it is empty on submit', () => {
				act(() => {
					fireEvent.input(contentTextarea, { target: { value: '' } });
					fireEvent.click(submitButton);
				});

				expect(mockedOnSubmit).not.toHaveBeenCalled();
				expect(container.textContent).toMatch(new RegExp('lessons.creation.form.fields.content.validation_rules.required'));
			});
		});

		describe('on change validation', () => {
			it('should trigger a "required" validation error on the content input on change after the first submit', () => {
				act(() => {
					fireEvent.input(contentTextarea, { target: { value: '' } });
					expect(container.textContent).not.toMatch(new RegExp('lessons.creation.form.fields.content.validation_rules.required'));
					fireEvent.click(submitButton);
					fireEvent.input(contentTextarea, { target: { value: 'a'.repeat(10) } });
					expect(container.textContent).not.toMatch(new RegExp('lessons.creation.form.fields.content.validation_rules.required'));

					fireEvent.input(contentTextarea, { target: { value: '' } });
				});

				expect(container.textContent).toMatch(new RegExp('lessons.creation.form.fields.content.validation_rules.required'));
			});
		});
	});
	*/

	describe('onSubmit', () => {
		it('should call the onSubmit prop method with the form\'s inputs\' values.', async () => {
			const mockedOnSubmit = jest.fn((data) => Promise.resolve(data));
			const { container } = render(<LessonCreationForm onSubmit={mockedOnSubmit} />);

			const expectedFormData = {
				title: 'dummy lesson title',
				description: 'dummy lesson description',
				content: '\\\n',
				// content: 'dummy lesson content',
			};

			await act(async () => {
				await fireEvent.input(container.querySelector('input[name="title"]'), { target: { value: expectedFormData.title } });
				await fireEvent.input(container.querySelector('textarea[name="description"]'), { target: { value: expectedFormData.description } });
				// await fireEvent.input(container.querySelector('textarea[name="content"]'), { target: { value: expectedFormData.content } });

				await fireEvent.click(container.querySelector('button[type="submit"]'));
			});

			expect(mockedOnSubmit).toHaveBeenCalledTimes(1);
			expect(mockedOnSubmit).toHaveBeenCalledWith(expectedFormData);
		});
	});
});
