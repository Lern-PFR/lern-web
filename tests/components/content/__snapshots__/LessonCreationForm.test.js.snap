// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<styled.form
  display="grid"
  gridGap="1em"
  height="max-content"
  onSubmit={[Function]}
>
  <LabeledInput
    defaultValue=""
    disabled={false}
    errorText=""
    hasError={false}
    hintText=""
    id="title"
    key="title"
    name="title"
    onBlur={[Function]}
    onChange={[Function]}
    placeholder="lessons.creation.form.fields.title.placeholder"
    textStyle="brevier"
    type="text"
  >
    lessons.creation.form.fields.title.label
  </LabeledInput>
  <LabeledTextArea
    defaultValue=""
    disabled={false}
    errorText=""
    hasError={false}
    hintText=""
    id="description"
    key="description"
    name="description"
    onBlur={[Function]}
    onChange={[Function]}
    placeholder="lessons.creation.form.fields.description.placeholder"
    rows={6}
    type="textarea"
  >
    lessons.creation.form.fields.description.label
  </LabeledTextArea>
  <LabeledTextArea
    defaultValue=""
    disabled={false}
    errorText=""
    hasError={false}
    hintText=""
    id="content"
    key="content"
    name="content"
    onBlur={[Function]}
    onChange={[Function]}
    rows={6}
    type="textarea"
  >
    lessons.creation.form.fields.content.label
  </LabeledTextArea>
  <PrimaryButton
    justifySelf="start"
    type="submit"
  >
    lessons.creation.form.action.submit
  </PrimaryButton>
</styled.form>
`;
